pkg/kubelet/kubelet_pods.go

```diff
@@ -885,13 +886,19 @@ func (kl *Kubelet) getPullSecretsForPod(pod *v1.Pod) []v1.Secret {
 		}
 		secret, err := kl.secretManager.GetSecret(pod.Namespace, secretRef.Name)
 		if err != nil {
-			klog.InfoS("Unable to retrieve pull secret, the image pull may not succeed.", "pod", klog.KObj(pod), "secret", klog.KObj(secret), "err", err)
+			klog.InfoS("Unable to retrieve pull secret, the image pull may not succeed.", "pod", klog.KObj(pod), "secret", klog.KObj(secret), "secretName", secretRef.Name, "err", err)
+			failedSecretNames = append(failedSecretNames, secretRef.Name)
 			continue
 		}
 
 		pullSecrets = append(pullSecrets, *secret)
 	}
 
+	if len(failedSecretNames) > 0 {
+		kl.recorder.Eventf(pod, v1.EventTypeWarning, "FailedToRetrieveImagePullSecret", 
```

Comment:
Events from the kubelet get throttled, and recording a bunch of events will prevent other possibly more important events from getting recorded, so ideally, we would only record this if the image pull actually fails (and maybe combine this with the event we're already recording in that case).

That would require plumbing the fact of the secret lookup failure through to where the image pull is actually done

Reply:
If i understand correctly, this PR will only introduce 1 extra warning message no matter how many secrets pull are failed. At this case, why would

Reply:
one event for every call to syncPod for every pod that has invalid secret references

syncPod is called frequently, and kubelets run many pods

Reply:
Thanks for giving more background context.

Right now I can think of:

1. Make the `getPullSecretsForPod()` also return the error log string, which includes an error of pulling secret failure.
2. pass that log into the callback, which actually pulls the image https://github.com/kubernetes/kubernetes/blob/24a71990e02edbfd0a05f4abfdedcab991525874/pkg/kubelet/kubelet.go#L1718 https://github.com/kubernetes/kubernetes/blob/a750d8054a6cb3167f495829ce3e77ab0ccca48e/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L891
3. Merge the error info into the existing info
https://github.com/kubernetes/kubernetes/blob/a750d8054a6cb3167f495829ce3e77ab0ccca48e/pkg/kubelet/kuberuntime/kuberuntime_container.go#L183 

The pros is the fix is kinda straight forward. I am not sure if it will cause any backward compatibility issue because it involved the changes in different layers.

Another approach I can think of is to create (or maybe there are some existing implementation) sth like a global key value store to store the log and another component can get the log message and combine it in the existing event, when the pull image steps are actually failed.

Please let me know your thought. Thanks !

Reply:
plumbing the pull secret lookup failure to the point where we record an event about the image pull failure makes the most sense to me

Reply:
Thanks again to explain more details in Slack. I will try to make the change asap this week.